Pre-built binaries, installer, and building from source guide (#1010)
NVIDIA/cuda-quantum: CMakeLists.txt

# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

> cmake_minimum_required(VERSION 3.26 FATAL_ERROR)

Reviewer: Is this necessary? The LLVM we're using only requires 3.13 and suggests 3.20.

Author: I can move this to a separate PR. 3.26 is necessary for the Python wheels. If specify a lower version here, then we need to make sure that we use the lower cmake version in the dev dependencies used to build the docker file, so that building with the lower cmake version is tested.

> cmake_minimum_required(VERSION 3.26 FATAL_ERROR)

Author: ```suggestion
cmake_minimum_required(VERSION 3.22 FATAL_ERROR)
```
Keeping at 3.22 for now, but this is entirely untested - which is not ideal.

  cmake_policy(SET CMP0074 NEW)
endif()

> if(POLICY CMP0135)

Reviewer: Maybe a comment on what this magic number, 135, means?

Author: done

option(CUDAQ_BUILD_TESTS "Build cudaq tests" ON)
option(CUDAQ_ENABLE_RPC_LOGGING "Enable verbose printout for client/server qpud connection." OFF)
option(CUDAQ_BUILD_RELOCATABLE_PACKAGE "Make CUDA Quantum install tree relocatable, system headers included." OFF)
option(CUDAQ_TEST_MOCK_SERVERS "Enable Remote QPU Tests via Mock Servers." OFF)
option(CUDAQ_DISABLE_RUNTIME "Build without the CUDA Quantum runtime, just the compiler toolchain." OFF)
option(CUDAQ_SKIP_MPI "Do not build with MPI, even when it is available." OFF)
option(CUDAQ_DISABLE_CPP_FRONTEND "Build without the CUDA Quantum C++ Clang-based Frontend." OFF)

> # We allow to pick up certain build configurations directly from the environment,
> # since this facilitates some of the packaging (e.g. python packages built based on the pyproject.toml).

Reviewer: ```suggestion
# Certain build configurations may be set directly in the environment.
# This facilitates some of the packaging (e.g. python packages built based on the pyproject.toml).
```

SET(CMAKE_INSTALL_RPATH "$ORIGIN:$ORIGIN/lib:$ORIGIN/lib/plugins:$ORIGIN/../lib:$ORIGIN/../lib/plugins")

SET(BLA_STATIC ON)
if(NOT DEFINED ZLIB_USE_STATIC_LIBS)
  SET(ZLIB_USE_STATIC_LIBS TRUE)
endif()
if(NOT DEFINED OPENSSL_USE_STATIC_LIBS)
  SET(OPENSSL_USE_STATIC_LIBS TRUE)
endif()

if(NOT DEFINED CUDAQ_ENABLE_STATIC_LINKING)
  SET(CUDAQ_ENABLE_STATIC_LINKING $ENV{CUDAQ_ENABLE_STATIC_LINKING})
endif()
if (CUDAQ_ENABLE_STATIC_LINKING)
  # Using -static-libgcc -static-libstdc++ offers more fine grained control here if needed.
  # At the time of writing, there wasn't any benefit to using those over -static.
  # However, while this so far seems to be working reasonably well, this will

>   # unfortunately also link glibc statically, which in general is likely to cause

Reviewer: Is this true? I'm sort of surprised that CMake would go through the hoops here, and do something that is arguably really dangerous/broken. https://news.ycombinator.com/item?id=23816748#:~:text=Yep.,excuses%20that%20make%20it%20impossible.

Author: I am not sure it is per se dangerous, but certainly won't work well in all cases, and it does produce an explicit warning that glibc still needs to be available (separately) at runtime. I think that a lot of the problems are primarily for building libraries that are then potentially later linked with other things, and maybe this configuration can work well to create a fully self-contained executable in some cases.

Reviewer: Here is a good answer to why one doesn't want to statically link glibc. "Dangerous" might not be the best adjective here, but it was used as a short hand for programs that can run fine, with weird, unexpected behavior, or arbitrary failures. Which is not ideal. 💥 

https://stackoverflow.com/questions/57476533/why-is-statically-linking-glibc-discouraged

Author: Agreed - I've come across and read the same thread. To be revised in a subsequent change.

  # issues sooner or later. A better option might be to include the required GNU C Library
  # and have the nvq++ compiler dynamically pick the newer one of the included or system one.
  # See also these posts:
  # - https://blog.gibson.sh/2017/11/26/creating-portable-linux-binaries/
  # - https://www.jertype.com/upgrading-glibc/
  # If the final application loads multiple dynamic libraries, and two or more
  # of those dynamic libraries are linked to different versions of the same static
  # libraries, then the different copies of those static libraries will conflict
  # with each other. If we keep linking everything statically, we should at least set 
  # CMAKE_<LANG>_VISIBILITY_PRESET to hidden and CMAKE_VISIBILITY_INLINES_HIDDEN to 1, 

>   # to avoid some of these issues.

Reviewer: The premise of all this is that `nvq++` *is* `clang++` under the hood and needs to be installed right along side of an identical `clang++`. That should be the default and fully supported. (But it is granted, that may also be problematic on some systems.)

Author: My concern here are the runtime libraries, which are linked with other user libraries (i.e. libraries that we don't build and have no control over) using nvq++ to build applications. My understanding based on what I read is that if our runtime library statically links glibc without hiding symbols as much as possible it can cause conflicts when it is then linked with a user library that may contain the same symbols (depending on how that library is built). I am not sure how common that would be.
